File: internal/database/sender_profiles.go
package database

import (
	"context"
	"database/sql"
	"encoding/json"
	"fmt"
	"time"
)

// GetSenderProfile returns a profile by type and identifier, or nil if not found
func (db *DB) GetSenderProfile(ctx context.Context, userID int64, profileType ProfileType, identifier string) (*SenderProfile, error) {
	query := `
		SELECT id, user_id, profile_type, identifier,
		       email_count, emails_archived, emails_notified,
		       slug_counts, label_counts, keyword_counts,
		       sender_type, summary,
		       first_seen_at, last_seen_at, modified_at, created_at
		FROM sender_profiles
		WHERE user_id = $1 AND profile_type = $2 AND identifier = $3
	`

	var p SenderProfile
	var slugCountsJSON, labelCountsJSON, keywordCountsJSON []byte

	err := db.conn.QueryRowContext(ctx, query, userID, profileType, identifier).Scan(
		&p.ID, &p.UserID, &p.ProfileType, &p.Identifier,
		&p.EmailCount, &p.EmailsArchived, &p.EmailsNotified,
		&slugCountsJSON, &labelCountsJSON, &keywordCountsJSON,
		&p.SenderType, &p.Summary,
		&p.FirstSeenAt, &p.LastSeenAt, &p.ModifiedAt, &p.CreatedAt,
	)
	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("failed to get sender profile: %w", err)
	}

	p.SlugCounts = make(map[string]int)
	p.LabelCounts = make(map[string]int)
	p.KeywordCounts = make(map[string]int)

	if err := json.Unmarshal(slugCountsJSON, &p.SlugCounts); err != nil {
		return nil, fmt.Errorf("failed to unmarshal slug_counts: %w", err)
	}
	if err := json.Unmarshal(labelCountsJSON, &p.LabelCounts); err != nil {
		return nil, fmt.Errorf("failed to unmarshal label_counts: %w", err)
	}
	if err := json.Unmarshal(keywordCountsJSON, &p.KeywordCounts); err != nil {
		return nil, fmt.Errorf("failed to unmarshal keyword_counts: %w", err)
	}

	return &p, nil
}

// UpsertSenderProfile creates or updates a sender profile
func (db *DB) UpsertSenderProfile(ctx context.Context, profile *SenderProfile) error {
	slugCountsJSON, err := json.Marshal(profile.SlugCounts)
	if err != nil {
		return fmt.Errorf("failed to marshal slug_counts: %w", err)
	}
	labelCountsJSON, err := json.Marshal(profile.LabelCounts)
	if err != nil {
		return fmt.Errorf("failed to marshal label_counts: %w", err)
	}
	keywordCountsJSON, err := json.Marshal(profile.KeywordCounts)
	if err != nil {
		return fmt.Errorf("failed to marshal keyword_counts: %w", err)
	}

	query := `
		INSERT INTO sender_profiles (
			user_id, profile_type, identifier,
			email_count, emails_archived, emails_notified,
			slug_counts, label_counts, keyword_counts,
			sender_type, summary,
			first_seen_at, last_seen_at, modified_at, created_at
		) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, NOW(), NOW())
		ON CONFLICT (user_id, profile_type, identifier)
		DO UPDATE SET
			email_count = EXCLUDED.email_count,
			emails_archived = EXCLUDED.emails_archived,
			emails_notified = EXCLUDED.emails_notified,
			slug_counts = EXCLUDED.slug_counts,
			label_counts = EXCLUDED.label_counts,
			keyword_counts = EXCLUDED.keyword_counts,
			sender_type = EXCLUDED.sender_type,
			summary = EXCLUDED.summary,
			last_seen_at = EXCLUDED.last_seen_at,
			modified_at = NOW()
	`

	_, err = db.conn.ExecContext(ctx, query,
		profile.UserID, profile.ProfileType, profile.Identifier,
		profile.EmailCount, profile.EmailsArchived, profile.EmailsNotified,
		slugCountsJSON, labelCountsJSON, keywordCountsJSON,
		profile.SenderType, profile.Summary,
		profile.FirstSeenAt, profile.LastSeenAt,
	)
	if err != nil {
		return fmt.Errorf("failed to upsert sender profile: %w", err)
	}

	return nil
}

// DeleteStaleProfiles removes profiles not modified in over 1 year
func (db *DB) DeleteStaleProfiles(ctx context.Context) (int64, error) {
	query := `DELETE FROM sender_profiles WHERE modified_at < NOW() - INTERVAL '1 year'`

	result, err := db.conn.ExecContext(ctx, query)
	if err != nil {
		return 0, fmt.Errorf("failed to delete stale profiles: %w", err)
	}

	return result.RowsAffected()
}

// GetHistoricalEmailsFromAddress returns the last N emails from a specific address
func (db *DB) GetHistoricalEmailsFromAddress(ctx context.Context, userID int64, address string, limit int) ([]*Email, error) {
	query := `
		SELECT id, user_id, from_address, subject, slug, keywords, summary,
		       labels_applied, bypassed_inbox, reasoning, COALESCE(human_feedback, ''), notification_sent, processed_at, created_at
		FROM emails
		WHERE user_id = $1 AND from_address = $2
		ORDER BY processed_at DESC
		LIMIT $3
	`
	return db.scanEmails(ctx, query, userID, address, limit)
}

// GetHistoricalEmailsFromDomain returns the last N emails from any address at a domain
func (db *DB) GetHistoricalEmailsFromDomain(ctx context.Context, userID int64, domain string, limit int) ([]*Email, error) {
	query := `
		SELECT id, user_id, from_address, subject, slug, keywords, summary,
		       labels_applied, bypassed_inbox, reasoning, COALESCE(human_feedback, ''), notification_sent, processed_at, created_at
		FROM emails
		WHERE user_id = $1 AND from_address LIKE '%@' || $2
		ORDER BY processed_at DESC
		LIMIT $3
	`
	return db.scanEmails(ctx, query, userID, domain, limit)
}

// scanEmails is a shared helper to scan email rows
func (db *DB) scanEmails(ctx context.Context, query string, args ...interface{}) ([]*Email, error) {
	rows, err := db.conn.QueryContext(ctx, query, args...)
	if err != nil {
		return nil, fmt.Errorf("failed to query emails: %w", err)
	}
	defer rows.Close()

	var emails []*Email
	for rows.Next() {
		var email Email
		var keywordsJSON, labelsJSON []byte

		err := rows.Scan(
			&email.ID, &email.UserID, &email.FromAddress, &email.Subject,
			&email.Slug, &keywordsJSON, &email.Summary,
			&labelsJSON, &email.BypassedInbox, &email.Reasoning,
			&email.HumanFeedback, &email.NotificationSent,
			&email.ProcessedAt, &email.CreatedAt,
		)
		if err != nil {
			return nil, fmt.Errorf("failed to scan email: %w", err)
		}

		if err := json.Unmarshal(keywordsJSON, &email.Keywords); err != nil {
			email.Keywords = []string{}
		}
		if err := json.Unmarshal(labelsJSON, &email.LabelsApplied); err != nil {
			email.LabelsApplied = []string{}
		}

		emails = append(emails, &email)
	}

	return emails, rows.Err()
}

// BuildProfileFromEmails computes counters from historical emails for bootstrapping
func BuildProfileFromEmails(userID int64, profileType ProfileType, identifier string, emails []*Email) *SenderProfile {
	profile := &SenderProfile{
		UserID:        userID,
		ProfileType:   profileType,
		Identifier:    identifier,
		EmailCount:    len(emails),
		SlugCounts:    make(map[string]int),
		LabelCounts:   make(map[string]int),
		KeywordCounts: make(map[string]int),
		FirstSeenAt:   time.Now(),
		LastSeenAt:    time.Now(),
	}

	for _, e := range emails {
		if e.Slug != "" {
			profile.SlugCounts[e.Slug]++
		}
		for _, label := range e.LabelsApplied {
			profile.LabelCounts[label]++
		}
		for _, kw := range e.Keywords {
			profile.KeywordCounts[kw]++
		}
		if e.BypassedInbox {
			profile.EmailsArchived++
		}
		if e.NotificationSent {
			profile.EmailsNotified++
		}
		if e.ProcessedAt.Before(profile.FirstSeenAt) {
			profile.FirstSeenAt = e.ProcessedAt
		}
		if e.ProcessedAt.After(profile.LastSeenAt) {
			profile.LastSeenAt = e.ProcessedAt
		}
	}

	return profile
}

File: cloudflare/src/db/sender-profiles.ts
import type { SenderProfile, SenderProfileRow, ProfileType, Email } from '../types/models';

function safeParseJSON<T>(text: string, fallback: T): T {
  try {
    return JSON.parse(text) as T;
  } catch {
    return fallback;
  }
}

function mapSenderProfile(row: SenderProfileRow): SenderProfile {
  return {
    id: row.id,
    userId: row.user_id,
    profileType: row.profile_type as ProfileType,
    identifier: row.identifier,
    emailCount: row.email_count,
    emailsArchived: row.emails_archived,
    emailsNotified: row.emails_notified,
    slugCounts: safeParseJSON<Record<string, number>>(row.slug_counts, {}),
    labelCounts: safeParseJSON<Record<string, number>>(row.label_counts, {}),
    keywordCounts: safeParseJSON<Record<string, number>>(row.keyword_counts, {}),
    senderType: row.sender_type,
    summary: row.summary,
    firstSeenAt: row.first_seen_at,
    lastSeenAt: row.last_seen_at,
    modifiedAt: row.modified_at,
    createdAt: row.created_at,
  };
}

const PROFILE_SELECT = `SELECT id, user_id, profile_type, identifier,
       email_count, emails_archived, emails_notified,
       slug_counts, label_counts, keyword_counts,
       sender_type, summary,
       first_seen_at, last_seen_at, modified_at, created_at
FROM sender_profiles`;

export async function getSenderProfile(
  db: D1Database,
  userId: number,
  profileType: ProfileType,
  identifier: string,
): Promise<SenderProfile | null> {
  const row = await db
    .prepare(`${PROFILE_SELECT} WHERE user_id = ? AND profile_type = ? AND identifier = ?`)
    .bind(userId, profileType, identifier)
    .first<SenderProfileRow>();
  return row ? mapSenderProfile(row) : null;
}

export async function getSenderProfileByID(
  db: D1Database,
  userId: number,
  profileId: number,
): Promise<SenderProfile | null> {
  const row = await db
    .prepare(`${PROFILE_SELECT} WHERE id = ? AND user_id = ?`)
    .bind(profileId, userId)
    .first<SenderProfileRow>();
  return row ? mapSenderProfile(row) : null;
}

export async function upsertSenderProfile(db: D1Database, profile: SenderProfile): Promise<void> {
  const slugCountsJSON = JSON.stringify(profile.slugCounts);
  const labelCountsJSON = JSON.stringify(profile.labelCounts);
  const keywordCountsJSON = JSON.stringify(profile.keywordCounts);

  await db
    .prepare(
      `INSERT INTO sender_profiles (
        user_id, profile_type, identifier,
        email_count, emails_archived, emails_notified,
        slug_counts, label_counts, keyword_counts,
        sender_type, summary,
        first_seen_at, last_seen_at, modified_at, created_at
      ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, datetime('now'), datetime('now'))
      ON CONFLICT (user_id, profile_type, identifier)
      DO UPDATE SET
        email_count = excluded.email_count,
        emails_archived = excluded.emails_archived,
        emails_notified = excluded.emails_notified,
        slug_counts = excluded.slug_counts,
        label_counts = excluded.label_counts,
        keyword_counts = excluded.keyword_counts,
        sender_type = excluded.sender_type,
        summary = excluded.summary,
        last_seen_at = excluded.last_seen_at,
        modified_at = datetime('now')`,
    )
    .bind(
      profile.userId,
      profile.profileType,
      profile.identifier,
      profile.emailCount,
      profile.emailsArchived,
      profile.emailsNotified,
      slugCountsJSON,
      labelCountsJSON,
      keywordCountsJSON,
      profile.senderType,
      profile.summary,
      profile.firstSeenAt,
      profile.lastSeenAt,
    )
    .run();
}

export async function deleteStaleProfiles(db: D1Database): Promise<number> {
  const result = await db
    .prepare("DELETE FROM sender_profiles WHERE modified_at < datetime('now', '-1 year')")
    .run();
  return result.meta.changes ?? 0;
}

// ============================================================================
// Utility functions (ported from Go models.go)
// ============================================================================

export const IGNORED_DOMAINS = new Set([
  'gmail.com', 'googlemail.com',
  'hotmail.com', 'outlook.com', 'live.com',
  'yahoo.com', 'yahoo.co.uk', 'aol.com',
  'icloud.com', 'me.com', 'mac.com',
  'protonmail.com', 'proton.me',
  'zoho.com', 'mail.com',
  'gmx.com', 'gmx.net',
  'yandex.com', 'tutanota.com', 'fastmail.com',
]);

export function isIgnoredDomain(domain: string): boolean {
  return IGNORED_DOMAINS.has(domain.toLowerCase());
}

export function extractDomain(email: string): string {
  const atIndex = email.lastIndexOf('@');
  if (atIndex < 0) return '';
  return email.slice(atIndex + 1).toLowerCase().replace(/[> ]+$/, '');
}

function topN(counts: Record<string, number>, n: number): string[] {
  return Object.entries(counts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, n)
    .map(([key, count]) => `${key} (${count})`);
}

export function topSlugs(profile: SenderProfile, n: number): string[] {
  return topN(profile.slugCounts, n);
}

export function topLabels(profile: SenderProfile, n: number): string[] {
  return topN(profile.labelCounts, n);
}

export function topKeywords(profile: SenderProfile, n: number): string[] {
  return topN(profile.keywordCounts, n);
}

export function bypassInboxRate(profile: SenderProfile): number {
  if (profile.emailCount === 0) return 0;
  return profile.emailsArchived / profile.emailCount;
}

export function notificationRate(profile: SenderProfile): number {
  if (profile.emailCount === 0) return 0;
  return profile.emailsNotified / profile.emailCount;
}

export function formatProfileForPrompt(profile: SenderProfile): string {
  const bRate = bypassInboxRate(profile) * 100;
  const nRate = notificationRate(profile) * 100;
  let result = `Type: ${profile.senderType} | Emails: ${profile.emailCount} | Archive rate: ${Math.round(bRate)}% | Notification rate: ${Math.round(nRate)}%\n`;

  const slugs = topSlugs(profile, 5);
  if (slugs.length > 0) {
    result += `Top slugs: ${slugs.join(', ')}\n`;
  }
  const labels = topLabels(profile, 5);
  if (labels.length > 0) {
    result += `Top labels: ${labels.join(', ')}\n`;
  }
  if (profile.summary) {
    result += `Summary: ${profile.summary}\n`;
  }
  return result;
}

/**
 * Builds a sender profile from historical emails (for bootstrapping).
 */
export function buildProfileFromEmails(
  userId: number,
  profileType: ProfileType,
  identifier: string,
  emails: Email[],
): SenderProfile {
  const now = new Date().toISOString();
  const profile: SenderProfile = {
    id: 0,
    userId,
    profileType,
    identifier,
    emailCount: emails.length,
    emailsArchived: 0,
    emailsNotified: 0,
    slugCounts: {},
    labelCounts: {},
    keywordCounts: {},
    senderType: '',
    summary: '',
    firstSeenAt: now,
    lastSeenAt: now,
    modifiedAt: now,
    createdAt: now,
  };

  for (const e of emails) {
    if (e.slug) {
      profile.slugCounts[e.slug] = (profile.slugCounts[e.slug] ?? 0) + 1;
    }
    for (const label of e.labelsApplied) {
      profile.labelCounts[label] = (profile.labelCounts[label] ?? 0) + 1;
    }
    for (const kw of e.keywords) {
      profile.keywordCounts[kw] = (profile.keywordCounts[kw] ?? 0) + 1;
    }
    if (e.bypassedInbox) profile.emailsArchived++;
    if (e.notificationSent) profile.emailsNotified++;
    if (e.processedAt < profile.firstSeenAt) profile.firstSeenAt = e.processedAt;
    if (e.processedAt > profile.lastSeenAt) profile.lastSeenAt = e.processedAt;
  }

  return profile;
}
